projects/RabbitMQ.Client/client/impl/AsyncConsumerWorkService.cs

```diff
@@ -83,9 +103,51 @@ async Task Loop()
                 }
             }
 
+            async Task LoopWithConcurrency(CancellationToken cancellationToken)
+            {
+                try
+                {
+                    while (await _channel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
+                    {
+                        while (_channel.Reader.TryRead(out Work work))
+                        {
+                            await _limiter.WaitAsync(cancellationToken);
```

Comment:
for a fast path this could be:
```csharp
// Do a quick synchronous check before we resort to async/await with the state-machine overhead.
if(!_limiter.Wait(0))
{
    await _limiter.WaitAsync(cancellationToken).ConfigureAwait(false);
}
```